lib/calendar.ts

```diff
@@ -1509,69 +1496,92 @@ const helperHebrew: HelperPerCalendarImpl = ObjectAssign({}, helperSharedImpl as
           }
         }
       }
-      return { ...calendarDate, day, month, monthCode, year, eraYear };
+      return { ...calendarDate, day, month, monthCode: monthCode as string, year, eraYear };
     }
-  },
+  }
   // All built-in calendars except Chinese/Dangi and Hebrew use an era
-  hasEra: false
-});
+  hasEra = false;
+}
 
 /**
  * For Temporal purposes, the Islamic calendar is simple because it's always the
  * same 12 months in the same order.
  */
-const helperIslamic: HelperPerCalendarImpl = ObjectAssign({}, helperSharedImpl, {
-  id: 'islamic',
-  calendarType: 'lunar',
-  inLeapYear(this: HelperPerCalendarImpl, calendarDate: CalendarYearOnly, cache: OneObjectCache) {
+abstract class IslamicBaseHelper extends HelperBase {
+  abstract id: BuiltinCalendarId;
+  calendarType = 'lunar' as const;
+  inLeapYear(calendarDate: CalendarYearOnly, cache: OneObjectCache) {
     // In leap years, the 12th month has 30 days. In non-leap years: 29.
     const days = this.daysInMonth({ year: calendarDate.year, month: 12, day: 1 }, cache);
     return days === 30;
-  },
+  }
   monthsInYear(/* calendarYear, cache */) {
     return 12;
-  },
-  minimumMonthLength: (/* calendarDate */) => 29,
-  maximumMonthLength: (/* calendarDate */) => 30,
-  DAYS_PER_ISLAMIC_YEAR: 354 + 11 / 30,
-  DAYS_PER_ISO_YEAR: 365.2425,
-  constantEra: 'ah',
-  estimateIsoDate(
-    this: HelperPerCalendarImpl & { DAYS_PER_ISLAMIC_YEAR: number; DAYS_PER_ISO_YEAR: number },
-    calendarDate: CalendarYMD
-  ) {
+  }
+  minimumMonthLength(/* calendarDate */) {
+    return 29;
+  }
+  maximumMonthLength(/* calendarDate */) {
+    return 30;
+  }
+  DAYS_PER_ISLAMIC_YEAR = 354 + 11 / 30;
+  DAYS_PER_ISO_YEAR = 365.2425;
+  constantEra = 'ah';
+  estimateIsoDate(calendarDate: CalendarYMD) {
     const { year } = this.adjustCalendarDate(calendarDate);
     return { year: MathFloor((year * this.DAYS_PER_ISLAMIC_YEAR) / this.DAYS_PER_ISO_YEAR) + 622, month: 1, day: 1 };
   }
-});
+}
 
-const helperPersian = ObjectAssign({}, helperSharedImpl, {
-  id: 'persian',
-  calendarType: 'solar',
-  inLeapYear(calendarDate: CalendarYearOnly, cache?: OneObjectCache) {
+// There are 6 Islamic calendars with the same implementation in this polyfill.
+// They vary only in their ID. They do emit different output from the underlying
+// Intl implementation, but our code for each of them is identical.
```

Comment:
I didn't realize this! I guess in a correct implementation they would have to have different Hijri correction days?

Reply:
I don't know exactly why they return different results, but they do return different results. My naive understanding of Islamic calendars is that the months change based on when the new moon happens, and "when the new moon happens" depends on who is doing the observing and where those observations are happening.

Furthermore, there is a "tabular" calendar which I believe means that they have some sort of mathematical formulae to determine the lunar cycles, vs. an actual observational mechanism to determine the month boundaries. https://en.wikipedia.org/wiki/Tabular_Islamic_calendar

For Temporal's purposes, I don't think we really care much about why these are different, only that they are different and we can't treat them all the same.